### PR TITLE
[feat]support ep_fsdp

### PR DESCRIPTION
**Pull request by kevssim**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

support ep_fsdp.

### Env&Config
NPU*8, Qwen3-30B-A3B, GBS=16, Grad_acc=4, ep_size=8, fsdp_size=8

### Loss curve comparison : ep_fsdp vs. pure fsdp
dlZhImageForImgTag

### Performance comparison

| Configuration | Speed (iter/s) | Memory (MB) | Speedup | Memory Reduction |
| :--- | :--- | :--- | :--- | :--- |
| ep_fsdp | 0.091 | 52488 | 1.40x | -8.6% |
| pure fsdp | 0.065 | 57403 | 1.00x | - |

### CODE REVIEW

File: src/twinkle/model/transformers/transformers.py [hunk 1]
```diff
@@ -219,13 +219,24 @@ def _decide_strategy(self, strategy: Literal['accelerate', 'native_fsdp']):
         self._enable_expert_parallel = self._should_enable_expert_parallel(self._expert_parallel_config,
                                                                            self.device_mesh)
         self._expert_parallel_applied = False
+        # Store ep_size for later use (EP mesh construction, grad clip, etc.)
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py]: 这里的逻辑是否可以封装进NativeFSDPStrategy，看没有其他地方使用

File: src/twinkle/model/transformers/transformers.py [hunk 2]
```diff
@@ -219,13 +219,24 @@ def _decide_strategy(self, strategy: Literal['accelerate', 'native_fsdp']):
...
         if use_native_fsdp:
+            ep_fsdp_mesh = None
+            if self._enable_expert_parallel and self.device_mesh is not None:
+                ep_fsdp_mesh = self.device_mesh.build_ep_fsdp_device_mesh(ep_size=self._ep_size)
             self.strategy = NativeFSDPStrategy(
                 mixed_precision=self.mixed_precision,
                 fsdp_config=self._fsdp_config,
                 device_mesh=self.device_mesh,
                 enable_ep=self._enable_expert_parallel,
+                ep_fsdp_device_mesh=ep_fsdp_mesh,
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py]: 这里是否可以在里面内部构造

File: src/twinkle/utils/device_mesh.py [hunk 3]
```diff
@@ -181,6 +181,42 @@ def get_dim_group(self, dims):
...
+
+    def build_ep_fsdp_device_mesh(self, ep_size: int = None):
+        import math
+        import torch
+        ep_size = ep_size or self.ep_size or 1
+        if ep_size <= 1:
+            return None
+        world_size = self.world_size
+        assert world_size % ep_size == 0, (f'world_size ({world_size}) must be divisible by ep_size ({ep_size})')
+        ep_fsdp_size = world_size // ep_size
```
Comment (tastelikefeet) [inline src/twinkle/utils/device_mesh.py]: 如果这里能算出来ep_fsdp_size，还需要外部传入吗？

File: src/twinkle/utils/device_mesh.py [hunk 4]
```diff
@@ -181,6 +181,42 @@ def get_dim_group(self, dims):
...
+        with torch.device('cpu'):
+            mesh = (
+                torch.arange(math.prod((ep_size, ep_fsdp_size)), dtype=torch.int).view(ep_fsdp_size,
+                                                                                       ep_size).transpose(0, 1))
+        return torch.distributed.DeviceMesh(self.device_type, mesh, mesh_dim_names=('ep', 'ep_fsdp'))
```
Comment (tastelikefeet) [inline src/twinkle/utils/device_mesh.py]: to_torch_device_mesh这个已经有了，考虑复用？

File: src/twinkle/utils/device_mesh.py [hunk 5]
```diff
@@ -118,7 +118,7 @@ def __post_init__(self):
         if not isinstance(self.mesh, np.ndarray):
             self.mesh = np.array(self.mesh)
 
-        valid_dim_names = {'dp', 'fsdp', 'tp', 'pp', 'cp', 'ep'}
+        valid_dim_names = {'dp', 'fsdp', 'tp', 'pp', 'cp', 'ep', 'ep_fsdp'}
```
Comment (tastelikefeet) [inline src/twinkle/utils/device_mesh.py:127]: 可以考虑使用from_sizes，这里应该就不用修改了

Comment (tastelikefeet) [reply src/twinkle/utils/device_mesh.py:127]: 参考下from_sizes的代码

File: src/twinkle/utils/device_mesh.py (hunk 3 above)
Comment (tastelikefeet) [reply src/twinkle/utils/device_mesh.py]: 或者说，这里如何判定需要在ep内部开启fsdp呢？

File: src/twinkle/utils/grad_clip.py [hunk 6]
```diff
@@ -85,11 +127,114 @@ def _local_grad(grad: torch.Tensor) -> torch.Tensor:
...
+        # non-EP: reduce over fsdp_group
+        non_ep_val = _local_norm_stat(non_ep_params, norm_type)
+        if fsdp_group is not None:
+            op = dist.ReduceOp.MAX if math.isinf(norm_type) else dist.ReduceOp.SUM
+            dist.all_reduce(non_ep_val, op=op, group=fsdp_group)
+
+        # EP: reduce over ep_fsdp_group, then ep_group
+        ep_val = _local_norm_stat(ep_params, norm_type)
+        if ep_fsdp_group is not None:
+            op = dist.ReduceOp.MAX if math.isinf(norm_type) else dist.ReduceOp.SUM
```
Comment (tastelikefeet) [inline src/twinkle/utils/grad_clip.py:174]: fsdp2是不是不需要使用all_reduce？

File: src/twinkle/utils/grad_clip.py [hunk 7]
```diff
@@ -74,7 +116,7 @@ def _local_grad(grad: torch.Tensor) -> torch.Tensor:
             local_norm = max(local_norm, local_grad.detach().abs().max().item())
         total_norm_tensor = torch.tensor(local_norm, device=reduce_device, dtype=torch.float32)
         if dist.is_initialized():
-            dist.all_reduce(total_norm_tensor, op=dist.ReduceOp.MAX, group=group)
+            dist.all_reduce(total_norm_tensor, op=dist.ReduceOp.MAX, group=reduce_group)
```
Comment (tastelikefeet) [inline src/twinkle/utils/grad_clip.py:119]: 同理，这里是否需要reduce

File: src/twinkle/model/transformers/transformers.py (hunk 1 above)
Comment (kevssim) [reply src/twinkle/model/transformers/transformers.py]: fixed

File: src/twinkle/model/transformers/transformers.py (hunk 2 above)
Comment (kevssim) [reply src/twinkle/model/transformers/transformers.py]: fixed

File: src/twinkle/utils/device_mesh.py (hunk 4 above)
Comment (kevssim) [reply src/twinkle/utils/device_mesh.py]: fixed

File: src/twinkle/utils/device_mesh.py (hunk 5 above)
Comment (kevssim) [reply src/twinkle/utils/device_mesh.py:127]: 这里不加'ep_fsdp'的话，构建DevicesMesh会assert失败

File: src/twinkle/utils/device_mesh.py (hunk 3 above)
Comment (kevssim) [reply src/twinkle/utils/device_mesh.py]: 目前的设计是：
当 ep_size > 1 时， 计算 ep_fsdp_size = world_size // ep_size。如果 ep_fsdp_size > 1，说明同一组 experts 分布在多个 rank 上，需要 FSDP 来管理专家参数的分片；如果 ep_fsdp_size == 1，专家层不需要额外 FSDP，但仍通过 torch.distributed.DeviceMesh 统一管理 EP 通信 group。

File: src/twinkle/utils/grad_clip.py (hunk 6 above)
Comment (kevssim) [reply src/twinkle/utils/grad_clip.py:174]: 这里的 all_reduce 不能去掉。
非 EP 参数和 EP 参数不在同一个mesh上,不能直接用torch.nn.utils.clip_grad_norm。所以这里先 to_local 提取各 rank 上的局部 norm 统计量，再按对应的 process group 做 all_reduce，最后才能得到正确的全局norm。

File: src/twinkle/utils/grad_clip.py (hunk 7 above)
Comment (kevssim) [reply src/twinkle/utils/grad_clip.py:119]: 这段代码只有当 parameters 列表中同时存在 DTensor 和普通 Tensor时才会走到，此时 torch.nn.utils.clip_grad_norm_ 无法统一处理，需要手动 to_local() 后计算范数再 all_reduce。

File: src/twinkle/utils/device_mesh.py (hunk 3 above)
Comment (tastelikefeet) [reply src/twinkle/utils/device_mesh.py]: 是不是可以从cookbook中直接传入？类似DeviceMesh.from_sizes(ep_size=, ep_fsdp_size=, ...) 你看下目前etp_size，应该就是这么处理的

Comment (kevssim) [reply src/twinkle/utils/device_mesh.py]: fixed

DeviceMesh：新增 ep_fsdp_size 字段（和 etp_size 同模式），from_sizes 在传入 ep_size > 1 时自动计算 ep_fsdp_size=world_size // ep_size，删除 build_ep_fsdp_device_mesh 方法。

cookbook：改用 DeviceMesh.from_sizes(fsdp_size=8, dp_size=1, ep_size=8)。
